src/zenml/materializers/built_in_materializer.py

```diff
@@ -66,11 +71,246 @@ def handle_input(self, data_type: Type[Any]) -> Any:
         return contents
 
     def handle_return(self, data: Any) -> None:
-        """Handles basic built-in types and stores them as json.
+        """Serialize a basic type to JSON.
 
         Args:
             data: The data to store.
         """
         super().handle_return(data)
-        filepath = os.path.join(self.artifact.uri, DEFAULT_FILENAME)
-        yaml_utils.write_json(filepath, data)
+        yaml_utils.write_json(self.data_path, data)
+
+
+class BytesMaterializer(BaseMaterializer):
+    """Handle `bytes` data type, which is not JSON serializable."""
+
+    ASSOCIATED_ARTIFACT_TYPES = (DataArtifact, DataAnalysisArtifact)
+    ASSOCIATED_TYPES = (bytes,)
+
+    def __init__(self, artifact: "BaseArtifact"):
+        """Define `self.data_path`.
+
+        Args:
+            artifact: Artifact required by `BaseMaterializer.__init__()`.
+        """
+        super().__init__(artifact)
+        self.data_path = os.path.join(self.artifact.uri, DEFAULT_BYTES_FILENAME)
+
+    def handle_input(self, data_type: Type[Any]) -> Any:
+        """Reads a bytes object from file.
+
+        Args:
+            data_type: The type of the data to read.
+
+        Returns:
+            The data read.
+        """
+        super().handle_input(data_type)
+        with open(self.data_path, "rb") as file_:
+            return file_.read()
+
+    def handle_return(self, data: Any) -> None:
+        """Save a bytes object to file.
+
+        Args:
+            data: The data to store.
+        """
+        super().handle_return(data)
+        with open(self.data_path, "wb") as file_:
+            file_.write(data)
+
+
+def _all_serializable(iterable: Iterable[Any]) -> bool:
+    """For an iterable, check whether all of its elements are JSON-serializable.
+
+    Args:
+        iterable (Iterable): The iterable to check.
+
+    Returns:
+        True if all elements are JSON-serializable, else False.
+    """
+    return all(_is_serializable(element) for element in iterable)
+
+
+def _is_serializable(obj: Any) -> bool:
+    """Check whether a built-in object is JSON-serializable.
+
+    Args:
+        obj: The object to check.
+
+    Returns:
+        True if the entire object is JSON-serializable, else False.
+    """
+    if any(isinstance(obj, basic_type) for basic_type in BASIC_TYPES):
+        return True
+    if any(isinstance(obj, iterable_) for iterable_ in (list, tuple)):
```

Comment:
Maybe include `set` here as well? Also, I think `isinstance(obj, ...)` allows a tuple as the second value to simplify the code a little

Reply:
Great suggestion, applied both!